### PR TITLE
Implements implicit `in world` loops

### PR DESCRIPTION
**Pull request by ike709**
Stuff like `for(var/mob/M)`

### CODE REVIEW

**Comment by wixoaGit** [comment]
These loops are not implicit `in world` loops. They loop through every existing instance of a type, wether or not it's in the world. For example, `for (var/client/C)` loops through every client, despite them not being in `world.contents`.

https://github.com/wixoaGit/OpenDream/issues/50#issuecomment-865178884

**Comment by ike709** [comment]
According to the docs, a for loop will by default loop through the world. Clients and datums are a special case and a separate issue.

image

**Comment by wixoaGit** [comment]
I suspect the docs aren't being entirely correct here. It works for _any_ type, there is no special case for just clients and datums.

I'd rather this not silently create a non-working loop for types that aren't atoms.

File: OpenDreamShared/Compiler/DM/DMParser.cs [hunk 1]
```diff
@@ -781,50 +781,66 @@ public partial class DMParser : Parser<Token> {
...
-                    DMASTProcBlockInner body = ProcBlock();
-                    if (body == null) {
-                        DMASTProcStatement statement = ProcStatement();
-
-                        if (statement == null) Error("Expected body or statement");
-                        body = new DMASTProcBlockInner(new DMASTProcStatement[] { statement });
+                    //Implicit "in world"
+                    if (rangeBegin is DMASTConstantNull && rangeEnd is null && defaultStep)
+                    {
+                        var path = new DreamPath(variableDeclaration.Type.ToString());
```
Comment (ike709) [inline OpenDreamShared/Compiler/DM/DMParser.cs:815]: Not sure why but the compiler hated it when I used `variableDeclaration.Type` directly.

File: DMCompiler/DM/Visitors/DMVisitorProcBuilder.cs [hunk 2]
```diff
@@ -212,7 +213,18 @@ class DMVisitorProcBuilder : DMASTVisitor {
                     _proc.BreakIfFalse();
 
                     DMASTProcStatementVarDeclaration varDeclaration = statementForList.Initializer as DMASTProcStatementVarDeclaration;
-                    if (varDeclaration != null && varDeclaration.Type != null) {
+                    if (varDeclaration != null && varDeclaration.Type != null)
+                    {
+
```
Comment (ike709) [inline DMCompiler/DM/Visitors/DMVisitorProcBuilder.cs:218]:
```suggestion
```

File: OpenDreamShared/Compiler/DM/DMParser.cs [hunk 3]
```diff
@@ -781,50 +781,60 @@ public partial class DMParser : Parser<Token> {
...
-                        body = new DMASTProcBlockInner(new DMASTProcStatement[] { statement });
-                    }
-
-                    return new DMASTProcStatementForStandard(initializer, comparator, incrementor, body);
+                    return new DMASTProcStatementForStandard(initializer, comparator, incrementor, GetForBody());
                 } else if (variableDeclaration != null) {
                     DMASTExpression rangeBegin = variableDeclaration.Value;
                     Whitespace();
-                    Consume(TokenType.DM_To, "Expected 'to'");
+                    if (rangeBegin is not DMASTConstantNull)
```
Comment (wixoaGit) [inline OpenDreamShared/Compiler/DM/DMParser.cs:788]: This allows you to do an implicit world loop if you do `for (var/atom/A = null)`. This does not work in byond.

File: OpenDreamShared/Compiler/DM/DMParser.cs [hunk 4]
```diff
@@ -781,50 +781,60 @@ public partial class DMParser : Parser<Token> {
...
+                    {
+                        Consume(TokenType.DM_To, "Expected 'to'");
+                    }
+
                     Whitespace();
                     DMASTExpression rangeEnd = Expression();
-                    if (rangeEnd == null) Error("Expected an expression");
+                    if (rangeBegin is not DMASTConstantNull && rangeEnd == null) Error("Expected an expression");
```
Comment (wixoaGit) [inline OpenDreamShared/Compiler/DM/DMParser.cs:795]: This would allow `for (var/i = null to)`

**Review by wixoaGit — CHANGES_REQUESTED** [review]

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**